src/governance.cpp

```diff
@@ -732,90 +725,105 @@ bool CGovernanceManager::ConfirmInventoryRequest(const CInv& inv)
 
-            // Push the inventory budget proposal message over to the other client
-            LogPrint("gobject", "CGovernanceManager::Sync -- syncing govobj: %s, peer=%d\n", strHash, pfrom->id);
-            pfrom->PushInventory(CInv(MSG_GOVERNANCE_OBJECT, it->first));
-            ++nObjCount;
+    if(netfulfilledman.HasFulfilledRequest(pnode->addr, NetMsgType::MNGOVERNANCESYNC)) {
+        // Asking for the whole list multiple times in a short period of time is no good
+        LogPrint("gobject", "CGovernanceManager::%s -- peer already asked me for the list\n", __func__);
+        Misbehaving(pnode->GetId(), 20);
+        return;
+    }
+    netfulfilledman.AddFulfilledRequest(pnode->addr, NetMsgType::MNGOVERNANCESYNC);
 
-            std::vector<CGovernanceVote> vecVotes = govobj.GetVoteFile().GetVotes();
-            for(size_t i = 0; i < vecVotes.size(); ++i) {
-                if(filter.contains(vecVotes[i].GetHash())) {
-                    continue;
-                }
-                if(!vecVotes[i].IsValid(true)) {
-                    continue;
-                }
-                pfrom->PushInventory(CInv(MSG_GOVERNANCE_OBJECT_VOTE, vecVotes[i].GetHash()));
-                ++nVoteCount;
-            }
+    int nObjCount = 0;
+    int nVoteCount = 0;
+
+    // SYNC GOVERNANCE OBJECTS WITH OTHER CLIENT
+
+    LogPrint("gobject", "CGovernanceManager::%s -- syncing all objects to peer=%d\n", __func__, pnode->id);
+
+    LOCK2(cs_main, cs);
+
+    // all valid objects, no votes
+    for(object_m_it it = mapObjects.begin(); it != mapObjects.end(); ++it) {
+        CGovernanceObject& govobj = it->second;
+        std::string strHash = it->first.ToString();
+
+        LogPrint("gobject", "CGovernanceManager::%s -- attempting to sync govobj: %s, peer=%d\n", __func__, strHash, pnode->id);
+
+        if(govobj.IsSetCachedDelete() || govobj.IsSetExpired()) {
+            LogPrintf("CGovernanceManager::%s -- not syncing deleted/expired govobj: %s, peer=%d\n", __func__,
+                      strHash, pnode->id);
+            continue;
         }
+
+        // Push the inventory budget proposal message over to the other client
+        LogPrint("gobject", "CGovernanceManager::%s -- syncing govobj: %s, peer=%d\n", __func__, strHash, pnode->id);
+        pnode->PushInventory(CInv(MSG_GOVERNANCE_OBJECT, it->first));
+        ++nObjCount;
     }
 
-    CNetMsgMaker msgMaker(pfrom->GetSendVersion());
-    connman.PushMessage(pfrom, msgMaker.Make(NetMsgType::SYNCSTATUSCOUNT, MASTERNODE_SYNC_GOVOBJ, nObjCount));
-    connman.PushMessage(pfrom, msgMaker.Make(NetMsgType::SYNCSTATUSCOUNT, MASTERNODE_SYNC_GOVOBJ_VOTE, nVoteCount));
-    LogPrintf("CGovernanceManager::Sync -- sent %d objects and %d votes to peer=%d\n", nObjCount, nVoteCount, pfrom->id);
+    CNetMsgMaker msgMaker(pnode->GetSendVersion());
+    connman.PushMessage(pnode, msgMaker.Make(NetMsgType::SYNCSTATUSCOUNT, MASTERNODE_SYNC_GOVOBJ, nObjCount));
```

Comment:
Not really related to this PR, but is the order in which SYNCSTATUSCOUNT and INV(MSG_GOVERNANCE_OBJECT) arrive in any way important? As I understand the code, there is no order guaranteed because inventories are sent/flushed in SendMessages, while SYNCSTATUSCOUNT is sent here immediately.

Reply:
`SYNCSTATUSCOUNT` is for informational purposes only, there is no logic tied to it, and yes, it doesn't matter in which order they arrive.